### PR TITLE
feat: add export of collections and recipes

### PR DESCRIPTION
**Pull request by adam-sroka**
This PR implements export functionality of Collections and Recipes via a) saving of csv files to the Downloads folder and b) of sharing csv files to other apps. Other minor refactoring improvements are also included.

### CODE REVIEW

File: packages/gibsonify_api/lib/src/models/recipe.dart [hunk 1]
```diff
@@ -109,6 +107,43 @@ class Recipe extends Equatable {
...
     return data;
   }
+
+  String toCsv() {
+    String recipeInfo = '"$employeeNumber","$number","$date","$name","$type",';
+    String csv = '';
+
+    String measurementsCombined = combineMeasurements(measurements);
+    var blankFieldsAfterMeasurements = ',,,,,,';
+    csv = csv +
```
Comment (adam-sroka) [inline packages/gibsonify_api/lib/src/models/recipe.dart:117]: all these `csv = csv + ...` cases could be handled by `+=`

File: dev_tools/gibsons_form_template.csv [hunk 2]
```diff
@@ -0,0 +1,6 @@
...
+2222,EmployeeNumber,ACDEFGH1234,Firstname Lastname,CC,+YY,12345678,YYYY-MM-DD,Normal day/Sick day/...,YYYY-MM-DD,HH:MM AM/PM,"Lat, Lon",Yes/No,NotCollectedReason,HH:MM AM/PM,Yes/No,YYYY-MM-DD,SecondVisitReason,Completed/Incomplete/...,NotCompletedReason,Comments,True/False,UUID4,FoodItemName4,Morning/Afternoon/...,Home made/Gift/...,IngredientDescriptionString,Raw/Roasted with oil/...,true/false,UUID6,YYYY-MM-DD,RecipeName,Method1_value1_unit1 + Method2_value2_unit2
+2222,EmployeeNumber,ACDEFGH1234,Firstname Lastname,CC,+YY,12345678,YYYY-MM-DD,Normal day/Sick day/...,YYYY-MM-DD,HH:MM AM/PM,"Lat, Lon",Yes/No,NotCollectedReason,HH:MM AM/PM,Yes/No,YYYY-MM-DD,SecondVisitReason,Completed/Incomplete/...,NotCompletedReason,Comments,True/False,UUID5,FoodItemName5,Morning/Afternoon/...,Home made/Gift/...,IngredientDescriptionString,Raw/Roasted with oil/...,true/false,UUID9,YYYY-MM-DD,RecipeName,Method1_value1_unit1 + Method2_value2_unit2
```
Comment (sshazril) [inline dev_tools/gibsons_form_template.csv:6]: same as above

**Review by sshazril — CHANGES_REQUESTED** [review]
Have reviewed most of the changes except the meat of the export logic. 
Main things to note is that 

- share via email functionality still occasionally doesn't work on my device - same error as previously shared (unable to attach files etc), small sissue since retrying sorts it + have not been able to reproduce reliably
- export recipes -> import recipes does not currently work, although I'm reasonably certain this is because recipe export exports recipe type as `Standard Recipe` instead of `Standard` as per the recipe template. Similarly measurement format is also currently incorrect
- other comments are included in review
Will review the main meat of the export bloc soon

File: lib/import_export/bloc/import_export_bloc.dart [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+  final GibsonifyRepository _gibsonifyRepository;
+
+  ImportExportBloc({required GibsonifyRepository gibsonifyRepository})
+      : _gibsonifyRepository = gibsonifyRepository,
+        super(const ImportExportState()) {
+    on<DataSavedToFiles>(_onDataSavedToFiles);
+    on<DataShared>(_onDataShared);
+  }
+
+  void _onDataSavedToFiles(
```
Comment (sshazril) [inline lib/import_export/bloc/import_export_bloc.dart:25]: Looks good for now, in the future we should probably separate exporting recipe data and collection data as two separate options, same for saving to internal memory etc

File: dev_tools/gibsons_form_example.csv [hunk 4]
```diff
@@ -0,0 +1,4 @@
...
+"2e25a839-f891-423b-a429-ce02d7e75200","1231231234","ABCDEFHG123","John Doe","IN","+91","8372837372","2022-03-28","Normal day","2022-03-30","11:20 AM","37.421998333333335, -122.084","No","it got lost","11:50 AM","No","2022-03-30","respondent had to pick up child","Incomplete","respondent had to go away","will attempt to collect data next time","true","05377a75-ce30-4d2e-abe0-5d27af7c617b","Banana","Evening (4:01 pm - 7:00 pm)","Purchased","banana","Raw","true","afe6342e-3029-4978-9288-8d1cd5b96f8c","2022-03-19","Banana","Size (photo)_Size number (photo)_5"
+"78fb9058-2952-4acc-84cd-a56b6483e932","1231231234","IJKFJLS13323","Peter Parker","DZ","+213","383883838","2022-03-28","Normal day","2022-03-30","2:39 PM","37.421998333333335, -122.084","Yes","null","5:41 PM","Yes","null","null","Completed","null","null","true","e941e66c-a9bd-4c65-bd31-a8489a434920","Strawberries","Night (7:01 pm - 3:00 am)","Home garden/farm","strawberry","Raw","true","dde03dc3-72bf-422a-8613-188c5152be2b","2022-03-19","Strawberry","Direct weight_Grams_250"
```
[the remark this answers is not in the document]
Comment (adam-sroka) [reply dev_tools/gibsons_form_example.csv:4]: Yup, it is that way in the UI but this somehow slipped past. Will do.

File: packages/gibsonify_api/lib/src/models/recipe.dart [hunk 5]
```diff
@@ -3,32 +3,30 @@ import 'package:uuid/uuid.dart';
...
-import 'recipe_probe.dart';
+import 'package:gibsonify_api/gibsonify_api.dart';
 
 class Recipe extends Equatable {
   Recipe({
-    this.recipeName,
+    this.name,
     this.employeeNumber,
-    String? recipeNumber,
+    String? number, // TODO: change this to `id` since its alphanumeric?
```
[the remark this answers is not in the document]
Comment (adam-sroka) [reply packages/gibsonify_api/lib/src/models/recipe.dart:12]: Yeah we can leave the TODO for later and then decide, I knew that that was the case with ICRISAT jargon, but strictly speaking it's not a number, but a string 😄!

File: pubspec.yaml [hunk 6]
```diff
@@ -27,6 +27,7 @@ dependencies:
   url_launcher: ^6.0.20
   file_picker: ^4.4.0
   csv: ^5.0.1
+  permission_handler: ^9.2.0
```
[the remark this answers is not in the document]
Comment (adam-sroka) [reply pubspec.yaml:30]: Previously, the data that was saved locally to device memory was saved inside the `Android` system directory (in the `org.gibsonify.gibsonify` subdirectory) which is normally not so accessible for users. In this PR the saving to the `Downloads` folder was implemented, which requires more permissions for the app. Therefore, before the saving of files is carried out, this package checks if the app has required permissions to save to the `Downloads` folder, and if it doesn't it requests them from the user.

File: lib/import_export/bloc/import_export_bloc.dart (hunk 3 above)
Comment (adam-sroka) [reply lib/import_export/bloc/import_export_bloc.dart:25]: Yup, I was thinking that we should also have options to export all Collections, not just the finished ones, but I wanted to stick to current feature requests. Currently, exporting everything together is the most straightforward (and hopefully also the least-bug-ridden) option, for example synchronizing the files to have the same timestamp in the name if they were saved asynchronously (and possibly by different blocs) would just be more cumbersome. I've done a couple of designs of the export while making this PR and for now, having everything bundled together in just one bloc a few event handlers seems the most straightforward, with possibility of extending and atomizing the functionality in the future.

**Comment by sshazril** [comment]
Changes all look good! I also tried the import -> export -> process and ran into a bug that was to do with line endings, it might be the cause of the issues you were having on your end too. Fortunately it's something I knew to look out for from previous experience, have just pushed a small commit to fix the bug. Import -> export -> import now works which is very exciting, hopefully it works on your end too. Else, LGTM - will be good to finally get this merged and sent to ICRISAT.

**Review by sshazril — APPROVED** [review]

**Comment by adam-sroka** [comment]
> Changes all look good! I also tried the import -> export -> process and ran into a bug that was to do with line endings, it might be the cause of the issues you were having on your end too. Fortunately it's something I knew to look out for from previous experience, have just pushed a small commit to fix the bug. Import -> export -> import now works which is very exciting, hopefully it works on your end too. Else, LGTM - will be good to finally get this merged and sent to ICRISAT.

Yes! It all works now, thank you so much 💙!

**Pull request merged**